pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,574 @@
+func (b *storageosUnmounter) TearDownAt(dir string) error {
+	if err := util.UnmountPath(dir, b.mounter); err != nil {
+		glog.V(4).Infof("Unmounted StorageOS volume %s failed with: %v", b.pvName, err)
+	}
+	if err := b.manager.UnmountVolume(b); err != nil {
+		glog.V(4).Infof("Mount reference for volume %s could not be removed from StorageOS: %v", b.pvName, err)
+	}
+	return nil
+}
+
+type storageosDeleter struct {
+	*storageos
+}
+
+var _ volume.Deleter = &storageosDeleter{}
+
+func (d *storageosDeleter) GetPath() string {
+	return getPath(d.podUID, d.volNamespace, d.volName, d.pvName, d.plugin.host)
+}
+
+func (d *storageosDeleter) Delete() error {
+	return d.manager.DeleteVolume(d)
+}
+
+type storageosProvisioner struct {
+	*storageos
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &storageosProvisioner{}
+
+func (c *storageosProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	// Apply ProvisionerParameters (case-insensitive). We leave validation of
+	// the values to the cloud provider.
+	for k, v := range c.options.Parameters {
+		switch strings.ToLower(k) {
+		case "description":
+			c.description = v
+		case "pool":
+			c.pool = v
+		case "fstype":
+			c.fsType = v
+		default:
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, c.plugin.GetPluginName())
+		}
+	}
+
+	// Set from PVC
+	c.volName = c.options.PVName
+	c.volNamespace = c.options.PVC.Namespace
+	c.labels = make(map[string]string)
+	for k, v := range c.options.PVC.Labels {
+		c.labels[k] = v
+	}
+	capacity := c.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	c.size = int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	vol, err := c.manager.CreateVolume(c)
+	if err != nil {
+		return nil, err
+	}
+	if vol.FSType == "" {
+		vol.FSType = defaultFSType
+	}
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   vol.Namespace + "~" + vol.Name,
+			Labels: map[string]string{},
+			Annotations: map[string]string{
+				volumehelper.VolumeDynamicallyCreatedByKey: "storageos-dynamic-provisioner",
+			},
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: c.options.PersistentVolumeReclaimPolicy,
+			AccessModes:                   c.options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceName(v1.ResourceStorage): resource.MustParse(fmt.Sprintf("%dGi", vol.Size)),
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				StorageOS: &v1.StorageOSVolumeSource{
+					VolumeName:      vol.Name,
+					VolumeNamespace: vol.Namespace,
+					Pool:            vol.Pool,
+					FSType:          vol.FSType,
+					ReadOnly:        false,
+				},
+			},
+		},
+	}
+	if len(c.options.PVC.Spec.AccessModes) == 0 {
+		pv.Spec.AccessModes = c.plugin.GetAccessModes()
+	}
+	if len(vol.Labels) != 0 {
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		for k, v := range vol.Labels {
```

Comment:
I believe the AWS, GCE, Flocker and Portworx provisioners also do this.  We think it's a useful feature as it allows k8s tools to query/verify storage parameters directly without having to use our api and tools.  

For example, you could query on PVs that have the production label where storageos.feature.replicas < 1 to show data at risk.